File: leetcode/57-insert-interval.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        if 0 == len(intervals):
            return [newInterval]

        merging = -1
        start = None

        ret = []
        for itv in intervals:
            if merging < 0:
                if newInterval.end < itv.start:
                    merging = 1
                    ret += [newInterval, itv]
                elif newInterval.end <= itv.end:
                    merging = 1
                    start = min(newInterval.start, itv.start)
                    ret += [Interval(start, itv.end)]
                elif newInterval.start <= itv.end:
                    merging = 0
                    start = min(newInterval.start, itv.start)
                else:
                    ret += [itv]
            elif 0 == merging:
                if newInterval.end < itv.start:
                    merging = 1
                    ret += [Interval(start, newInterval.end), itv]
                elif newInterval.end <= itv.end:
                    merging = 1
                    ret += [Interval(start, itv.end)]
            else: # merging > 0
                ret += [itv]

        if merging < 0:
            ret += [newInterval]
        elif 0 == merging:
            ret += [Interval(start, newInterval.end)]

        return ret

File: leetcode/57-insert-interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
 public:
  vector<Interval> insert(vector<Interval>& intervals,
                          Interval newInterval)
  {
    vector<Interval> ret;

    if (intervals.empty()) {
      ret.push_back(newInterval);
      return ret;
    }

    int start;
    int merging = -1;
    for (int i = 0; i < intervals.size(); ++i) {
      Interval& interval = intervals[i];
      switch (merging) {
        case -1:
          if (newInterval.end < interval.start) {
            merging = 1;
            ret.push_back(newInterval);
            ret.push_back(interval);
          } else if (newInterval.end <= interval.end) {
            merging = 1;
            start = min(interval.start, newInterval.start);
            ret.push_back(Interval(start, interval.end));
          } else if (newInterval.start <= interval.end) {
            merging = 0;
            start = min(interval.start, newInterval.start);
          } else {
            ret.push_back(interval);
          }
          break;
        case 0:
          if (newInterval.end < interval.start) {
            merging = 1;
            ret.push_back(Interval(start, newInterval.end));
            ret.push_back(interval);
          } else if (newInterval.end <= interval.end) {
            merging = 1;
            ret.push_back(Interval(start, interval.end));
          }
          break;
        case 1:
          ret.push_back(interval);
          break;
        default: /* empty */
          break;
      }
    }

    if (-1 == merging)
      ret.push_back(newInterval);
    else if (0 == merging)
      ret.push_back(Interval(start, newInterval.end));

    return ret;
  }
};
